camel/utils/channel.py

```diff
@@ -0,0 +1,67 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+import asyncio
+import uuid
+
+
+class AsyncSafeDict:
+    def __init__(self):
+        self.dict = {}
+        self.lock = asyncio.Lock()
+
+    async def put(self, key, value):
+        async with self.lock:
+            self.dict[key] = value
+
+    async def get(self, key, default=None):
+        async with self.lock:
+            return self.dict.get(key, default)
+
+    async def pop(self, key, default=None):
+        async with self.lock:
+            return self.dict.pop(key, default)
+
+    async def keys(self):
+        async with self.lock:
+            return list(self.dict.keys())
+
+
+class Channel:
+    def __init__(self):
+        self.receive_queue = asyncio.Queue()  # save received messgae
+        self.send_dict = AsyncSafeDict()
+
+    async def receive_from(self):
+        message = await self.receive_queue.get()
```

Comment:
In the example, all units share one channel and all try to get message out of the queue. If the workforce gets the message out of the queue which is not matched with the workforce id. Does it mean the queue message is gone without processed?

Reply:
Yes, so from my dessign, every subtask should be assigned to a workforce as it is produced. Every workforce will only process tasks assign to itself.